core/include/detray/propagator/actors/perigee_stopper.hpp

```diff
@@ -0,0 +1,125 @@
+ */
+
+#pragma once
+
+// Project include(s)
+#include "detray/definitions/algebra.hpp"
+#include "detray/definitions/detail/qualifiers.hpp"
+#include "detray/definitions/geometry.hpp"
+#include "detray/geometry/mask.hpp"
+#include "detray/geometry/shapes/line.hpp"
+#include "detray/geometry/tracking_volume.hpp"
+#include "detray/navigation/intersection/ray_intersector.hpp"
+#include "detray/navigation/navigator.hpp"
+#include "detray/propagator/base_actor.hpp"
+#include "detray/propagator/constrained_step.hpp"
+#include "detray/tracks/ray.hpp"
+#include "detray/utils/curvilinear_frame.hpp"
+
+namespace detray {
+
+/// Actor that exits from a navigation stream, if it has reached the perigee
+/// @Note Currently only works with step constraints (will be changed in the
+/// future)
+template <concepts::algebra algebra_t>
+struct perigee_stopper : actor {
+
+    using scalar_t = dscalar<algebra_t>;
+
+    struct state {
+        /// Radius around the beamline, where to test against the perigee
+        /// Outside this radius, the track is considered as not originating
+        /// from the IP
+        scalar_t m_stopping_radius{10.f * unit<scalar_t>::mm};
+        /// Tolerance under which to consider the track at the perigee
+        // @TODO Make smaller once overstepping is solved
+        scalar_t m_on_perigee_tol{100.f * unit<scalar_t>::um};
+        /// Index of the innermost volume for this detector: Convention is 0
+        unsigned int m_inner_vol_idx{0u};
+    };
+
+    /// Intersects a linear track approximation with the perigee and exits
+    /// the navigation, if the perigee is reached.
+    ///
+    /// @param prop_state state of the propagation
+    template <typename propagator_state_t>
+    DETRAY_HOST_DEVICE void operator()(state &actor_state,
+                                       propagator_state_t &prop_state) const {
+        using detector_t = typename propagator_state_t::detector_type;
+        using perigee_intersector_t = ray_intersector<line_circular, algebra_t>;
+
+        // Nothing left to do. Propagation will exit successfully on its own
+        auto &navigation = prop_state._navigation;
+        if (navigation.is_complete()) {
+            return;
+        }
+
+        // Only check this in the innermost volume and during backward nav.
+        if (navigation.volume() != actor_state.m_inner_vol_idx ||
+            navigation.direction() != navigation::direction::e_backward) {
+            return;
+        }
+
+        // Volume that contains the IP
+        const tracking_volume inner_vol{navigation.detector(),
+                                        actor_state.m_inner_vol_idx};
+
+        // Stop at the perigee (cylindrical detectors only)
+        if (inner_vol.id() != volume_id::e_cylinder) {
+            return;
+        }
+
+        auto &stepping = prop_state._stepping;
+        auto &track = stepping();
+
+        // Linear track approximation
+        const detail::ray<algebra_t> trk_approx{track.pos(),
```

Comment:
Why using linear approximation? This will drop precision for low pT particle

Reply:
Ray surface intersections are guaranteed to produce a solution if there is an interseciton. The line-mask is set large enough so that it should capture all relevant tracks to stop them (at least so far, the navigation rays have not been off by 1cm). If it does not work, I will come back and fix that in a future PR